Source/inv.cpp

```diff
@@ -1109,6 +1109,70 @@ void CleanupItems(Item *item, int ii)
 	}
 }
 
+enum can_put_result : uint8_t {
+	// clang-format off
+	CANPUT_NO_ADJACENT_ITEMS,
+	CANPUT_FAIL,
+	CANPUT_OK,
+	// clang-format on
+};
+
+can_put_result CanPutAdjacent(Point position)
+{
+	if (!CanPut(position))
+		return CANPUT_FAIL;
+
+	// check adjacent tiles to avoid dropping items in unreachable places
+	for (int x = -1; x <= 1; x++) {
+		for (int y = -1; y <= 1; y++) {
+			Point pos = position + Displacement { x, y };
+			if (InDungeonBounds(pos)) {
+				if (dItem[pos.x][pos.y] != 0)
+					return CANPUT_OK;
+			}
+		}
+	}
+
+	return CANPUT_NO_ADJACENT_ITEMS;
+}
+
+std::optional<Point> FindClosestValidItemPosition(Point startingPosition, unsigned int minimumRadius, unsigned int maximumRadius)
+{
+	if (minimumRadius > maximumRadius) {
+		return {}; // No valid search space with the given params.
+	}
+
+	if (minimumRadius == 0U) {
+		if (CanPut(startingPosition)) {
+			return startingPosition;
+		}
+		minimumRadius = 1;
+	}
+
+	for (int range = minimumRadius; range <= maximumRadius; range++) {
+		int adjacentFails = 0;
+		for (int yy = -range; yy <= range; yy++) {
+			// optimization to only check edges of the square
+			// makes sure we don't check same tiles multiple times
+			int step = range * 2;
+			if (yy == -range || yy == range)
+				step = 1;
+			for (int xx = -range; xx <= range; xx += step) {
+				Point pos = startingPosition + Displacement { xx, yy };
+				can_put_result res = CanPutAdjacent(pos);
+				if (res == CANPUT_OK)
+					return pos;
+				if (res == CANPUT_NO_ADJACENT_ITEMS)
+					adjacentFails++;
+			}
+		}
+		if (adjacentFails != 0)
+			break;

```

Comment:
```suggestion
		bool continueFloodFill = false;
		for (int yy = -range; yy <= range; yy++) {
			// optimization to only check edges of the square
			// makes sure we don't check same tiles multiple times
			int step = range * 2;
			if (yy == -range || yy == range)
				step = 1;
			for (int xx = -range; xx <= range; xx += step) {
				Point pos = startingPosition + Displacement { xx, yy };
				can_put_result res = CanPutAdjacent(pos);
				if (res == CANPUT_OK)
					return pos;
				if (dItem[pos.x][pos.y] != 0)
					continueFloodFill = true;
			}
		}
		if (!continueFloodFill)
			break;
```
This allows items to continue to be dropped as long as the adjacency check has a chance to succeed on the next largest radius. Also lets you remove the can_put_result enum and have CanPutAdjacent return true/false.